File: src/Nebula/Database/Models/Receivable.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using Nebula.Common.Models;

namespace Nebula.Database.Models;

/// <summary>
/// Cuentas por Cobrar.
/// </summary>
[BsonIgnoreExtraElements]
public class Receivable : IGenericModel
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;

    /// <summary>
    /// Tipo Operación: 'CARGO' | 'ABONO'
    /// </summary>
    public string Type { get; set; } = "CARGO";

    /// <summary>
    /// Identificador del Contacto.
    /// </summary>
    public string ContactId { get; set; } = string.Empty;

    /// <summary>
    /// Nombre de Contacto.
    /// </summary>
    public string ContactName { get; set; } = string.Empty;

    /// <summary>
    /// Observación / Comentario.
    /// </summary>
    public string Remark { get; set; } = string.Empty;

    /// <summary>
    /// Clave foránea comprobante de venta.
    /// </summary>
    public string InvoiceSale { get; set; } = "-";

    /// <summary>
    /// Configura el Tipo de comprobante.
    /// </summary>
    public string DocType { get; set; } = "-";

    /// <summary>
    /// Documento de Venta: F001-1.
    /// </summary>
    public string Document { get; set; } = string.Empty;

    /// <summary>
    /// Forma de Pago: Yape | Depósito | Contado.
    /// </summary>
    public string FormaPago { get; set; } = string.Empty;

    /// <summary>
    /// Monto de la deuda.
    /// </summary>
    public decimal Cargo { get; set; }

    /// <summary>
    /// Monto del Pago de la deuda.
    /// </summary>
    public decimal Abono { get; set; }

    /// <summary>
    /// Saldo pendiente a Cobrar.
    /// </summary>
    [BsonIgnore]
    public decimal Saldo { get; set; }

    /// <summary>
    /// Estado del Cargo: 'PENDIENTE' | 'COBRADO' | '-'.
    /// </summary>
    public string Status { get; set; } = "PENDIENTE";

    /// <summary>
    /// Identificador del Terminal de Venta.
    /// </summary>
    public string CajaDiaria { get; set; } = "-";

    /// <summary>
    /// Nombre del Terminal de Venta.
    /// </summary>
    public string Terminal { get; set; } = string.Empty;

    /// <summary>
    /// Identificador del Cargo.
    /// </summary>
    public string ReceivableId { get; set; } = "-";

    /// <summary>
    /// Fecha de Operación.
    /// </summary>
    public string CreatedAt { get; set; } = DateTime.Now.ToString("yyyy-MM-dd");

    /// <summary>
    /// Fecha de Vencimiento a Pagar.
    /// </summary>
    public string EndDate { get; set; } = "-";

    /// <summary>
    /// Año de registro.
    /// </summary>
    public string Year { get; set; } = DateTime.Now.ToString("yyyy");

    /// <summary>
    /// Mes de registro.
    /// </summary>
    public string Month { get; set; } = DateTime.Now.ToString("MM");
}

File: src/angular/src/app/receivable/interfaces/receivable.ts
import moment from "moment";

export type TypeReceivable = "CARGO" | "ABONO";
export type StatusReceivable = "PENDIENTE" | "COBRADO" | "-";

export class Receivable {
  constructor(
    public id: string = "-",
    public companyId: string = "",
    public type: TypeReceivable = "CARGO",
    public contactId: string = "",
    public contactName: string = "",
    public remark: string = "",
    public invoiceSale: string = "-",
    public docType: string = "-",
    public document: string = "-",
    public formaPago: string = "-",
    public cargo: number = 0,
    public abono: number = 0,
    public saldo: number = 0,
    public status: StatusReceivable = "PENDIENTE",
    public cajaDiaria: string = "-",
    public terminal: string = "-",
    public receivableId: string = "-",
    public createdAt: string = moment().format("YYYY-MM-DD"),
    public endDate: string = "",
    public year: string = moment().format("YYYY"),
    public month: string = moment().format("MM")) {
  }
}

export class ReceivableDataModal {
  constructor(
    public title: string = "",
    public type: "ADD" | "EDIT" = "ADD",
    public receivable: Receivable = new Receivable()) {
  }
}

export class ReceivableDetailDataModal {
  constructor(
    public cargo: Receivable = new Receivable(),
    public abonos: Array<Receivable> = new Array<Receivable>()) {
  }
}

export class ReceivableRequestParams {
  constructor(
    public year: string = "",
    public month: string = "",
    public status: string = "-") {
  }
}

export class CuentaPorCobrarParam {
  status: string = "";
  year: string = "";
}

export class CuentaPorCobrarClienteAnualParam extends CuentaPorCobrarParam {
  contactId: string = "";
}

export class CuentaPorCobrarMensualParam extends CuentaPorCobrarParam {
  month: string = "";
}

export class DeudaCliente {
  contactId: string = '';
  contactName: string = '';
  deudaTotal: number = 0;
  receivables: Receivable[] = [];
}

export class ResumenDeuda {
  totalPorCobrar: number = 0;
  deudasPorCliente: DeudaCliente[] = [];
}
